ext/filter/logical_filters.c

```diff
@@ -796,35 +796,70 @@ static int _php_filter_validate_ipv6(char *str, size_t str_len) /* {{{ */
 				return 0;
 			}
 			if (*str == ':') {
-				if (compressed) {
+				if (compressed_pos >= 0) {
 					return 0;
 				}
-				blocks++; /* :: means 1 or more 16-bit 0 blocks */
-				compressed = 1;
-
+				if (ip && blocks < 8) {
+					ip[blocks] = -1;
+				}
+				compressed_pos = blocks++; /* :: means 1 or more 16-bit 0 blocks */
 				if (++str == end) {
-					return (blocks <= 8);
+					if (blocks > 8) {
+						return 0;
+					} else {
+						goto fixup_ip;
+					}
 				}
 			} else if ((str - 1) == s) {
 				/* don't allow leading : without another : following */
 				return 0;
 			}
 		}
-		n = 0;
+		num = n = 0;
 		while ((str < end) &&
 		       ((*str >= '0' && *str <= '9') ||
 		        (*str >= 'a' && *str <= 'f') ||
 		        (*str >= 'A' && *str <= 'F'))) {
+			if (ip) {
+				if (*str >= '0' && *str <= '9') {
+					num = 16 * num + (*str - '0');
+				} else if (*str >= 'a' && *str <= 'f') {
+					num = 16 * num + (*str - 'a') + 10;
+				} else if (*str >= 'A' && *str <= 'F') {
+					num = 16 * num + (*str - 'A') + 10;
+				}
+			}
 			n++;
 			str++;
 		}
+		if (ip && blocks < 8) {
+			ip[blocks] = num;
+		}
 		if (n < 1 || n > 4) {
 			return 0;
 		}
 		if (++blocks > 8)
 			return 0;
 	}
-	return ((compressed && blocks <= 8) || blocks == 8);
+
+fixup_ip:
+	if (ip && ipv4) {
+		for (i = 0; i < 5; i++) {
+			ip[i] = 0;
+		}
+		ip[i++] = 0xffff;
+		ip[i++] = 256 * ip4elm[0] + ip4elm[1];
+		ip[i++] = 256 * ip4elm[2] + ip4elm[3];
+	} else if (ip && compressed_pos >= 0 && blocks < 8) {
+		for (i = compressed_pos + 1, j = 7; i < blocks; i++, j--) {
+			ip[j] = ip[i];
```

Comment:
Doesn't this reverse the order of the non-compressed blocks?

Reply:
Oops, I've totally screwed that up. Thank you for noticing!